File: app/src/main/java/br/com/nicetuber/feature/home/HomeInteractor.java
package br.com.nicetuber.feature.home;

import com.annimon.stream.Stream;

import java.util.List;
import java.util.Locale;

import javax.inject.Inject;

import br.com.nicetuber.base.BaseInteractor;
import br.com.nicetuber.model.ChannelSearched;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by thiagozg on 15/10/2017.
 */

public class HomeInteractor extends BaseInteractor {

    @Inject
    public HomeInteractor() {
    }

    public void searchChannel(IHome.Callback callback, String query) {
        youtubeApi.searchChannel(query, Locale.getDefault().getCountry())
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(
                    response -> {
                        List<ChannelSearched> listChannelsFiltered =
                                Stream.of(response.getListChannelSearcheds())
                                        .filter(channel -> !channel.getSnippet().getDescription().isEmpty())
                                        .toList();
                        callback.onSearchChannelSuccess(listChannelsFiltered);
                    },
                    throwable -> {
                        callback.onSearchChannelError(throwable.getMessage());
                        throwable.printStackTrace();
                    }
            );
    }
}

File: app/src/main/java/br/com/nicetuber/feature/home/HomeInteractor.kt
package br.com.nicetuber.feature.home

import br.com.nicetuber.base.BaseInteractor
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import java.util.*
import javax.inject.Inject

/**
 * Created by thiagozg on 15/10/2017.
 */

class HomeInteractor @Inject
constructor() : BaseInteractor() {

    fun searchChannel(callback: IHome.Callback, query: String) {
        youtubeApi.searchChannel(query, Locale.getDefault().country)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        { response ->
                            val listChannelsFiltered = response.listChannelSearcheds!!
                                    .filter { channel -> !channel.snippet!!.description!!.isEmpty() }
                                    .toList()
                            callback.onSearchChannelSuccess(listChannelsFiltered)
                        } , { throwable ->
                            callback.onSearchChannelError(throwable.message!!)
                            throwable.printStackTrace()
                        })
    }
}
